### ISSUE TITLE
ocamlfind for dedukti ?

### ISSUE THREAD

**Issue by Gbury** [Deducteam/Dedukti#110]
Is there a dedukti equivalent of ocamlfind for dedukti ?

More precisely, suppose some project wants to provide dedukti files. Then it will most probably install these dedukti files (possibly compiled) in its shared directory (or really any dedicated directory). The problem is afterwards, how can a lambda user make use of these installed files ? Currently, the only way I can see is for the user to call `dkcheck -I /path/to/share/dir/ file.dk` which is particularly annoying/difficult in practice, especially if you want to use more than one such library/project.
Imagine an opam package `foo` install these files in a subdirectory of its share directory (because it has other shared files to install), then you would have to call

`dkcheck -I ~/.opam/current_switch/share/foo/dkfiles file.dk`

which is lengthy, requires to be aware of the current opam switch, and requires the user to know precisely where `foo` installed the dedukti files. A somewhat more advanced opam user may do 

```dkcheck -I `opam config var foo:share`/dkfiles file.dk```

but it still seems way overcomplicated.

Is there any plan to ease this kind of use ? Currently, sharing files between dedukti developments (or even generating dedukti proofs which import from some static files), is ~~incredibly headache-inducing~~ harder than it should be, akin to trying to build ocaml code without ocamlfind.

Ideally, the call to dkcheck would look like: `dkcheck -package foo file.dk` (or `dkfind dkcheck -package foo file.p` if it emulates ocamlfind...).

P.S: This issue is actually quite closely related to namespaces, so the solution should probably introduce a notion of file/module hierarchy if dedukti wants to avoid the problem that ocaml has currently.

**Comment by francoisthire** [comment]
We had already some discussion about this but nothing real. Right now, my
opinion is that `dkcheck` should work the same way as `ocamlc`. If you want
a better way to handle dependencies, maybe we should enhance `dkdep` or
invent (again) another tool.

However, I am not sure to understand why we need to introduce a file/module
hierarchy and what kind of problem you want to avoid.

Le mer. 29 août 2018 à 21:28, Guillaume Bury <notifications@github.com> a
écrit :

**Comment by Gbury** [comment]
Suppose you have two libraries `A`and `B`, with `A` which depends on `B`. Now suppose both want to define a module named `basic.dk` or `misc.dk`, that would cause conflicts, and prevent `B` from depending on `A`. Now suppose `B` and `A` are independent and both define `basic.dk`, and the user in its development of project `C` wants to depend on both `A` and `B`, well it can't... That's a problem.
In ocaml, that means that projects that very usual file name such as `misc.ml`, `util.ml`, or `debug.ml` are problematic, and this is why packs were introduced, and lately why prefixing module name has become standard.

Now, ocaml did go the way of having external tools, but as people are realizing now, it's not really a good option: most ocaml package today either pack modules in a single one (which is hard to compile correctly, produce big non-modular files and binaries, etc...), or prefix modules with the name of the library (so have files `foo__misc.dk`) and then use module aliases to use saner names (such as `Foo.Misc`). But the consensus is that this situation is not really great and that namespace, i.e having package names so that a module is identified by a path package/module, would be better. However, that requires compiler support to implement properly.

So basically, the point is that avoiding name clashes is better implemented as part of the core compiler rather than external tools (because external tools will need to rename things, and that's messy).

**Comment by francoisthire** [comment]
What do you think about this @rafoo @Gaspi ?

**Comment by rafoo** [comment]
I think there are two questions here:
1) How to locate an installed Dedukti library?
2) How to compile a Dedukti development with dependencies?

For the second question, the `-I` solution has always been sufficient for me. For the first question however, I am unhappy with the solution I used in Dktactics:
- `make install` puts in PATH a one-line script called `dktactics` that simply outputs the install location of the library;
- developments using dktactics are compiled with `dkcheck -I $(dktactics)`.

I find this very hackich and I dislike to pollute PATH this way.


So I think the first feature of ocamlfind that we need for Dedukti is the handling of a database of locations for installed libraries (the `query`, `install`, and `remove` subcommands of ocamlfind).

**Comment by francoisthire** [comment]
This might be interesting to discuss during the next Dedukti meeting. This
would join an idea raised by Gilles that a Dedukti file could reference an
external source such as a URL.

**Comment by Gbury** [comment]
@rafoo what about the file name collision problem ?

**Comment by fblanqui** [comment]
Why not using opam?


Le 04/09/2018 à 11:53, Raphaël Cauderlier a écrit :

**Comment by Gbury** [comment]
opam may serve as a package manager to distribute dedukti packages, but by itself, it cannot handle the problem of specifying dependencies on the command line when compiling dedukti files, so there must be some support baked in dedukti.

**Comment by fblanqui** [comment]
Because OCaml and Coq users are already using opam and it is not a good 
idea to write a tool like opam from scratch.


Le 04/09/2018 à 13:21, fthire a écrit :

**Comment by fblanqui** [comment]
Right. @rlepigre proposes to use compiler options "-I path" and recursively qualified names (e.g. M.N.P denotes file M/N/P.dk) to solve this problem.

**Comment by francoisthire** [comment]
So a proposition would be the one of @rafoo: A command `dkfind` with three subcommands `query` `install` and `remove`. The packages would be installed with `opam` so that `dkfind` knows where to look for a package.

However, this does not solve the aliasing issue. Should we add another layer to handle packages?  Should we allow an arbitrary number of layers? If so, what would be their meaning?

**Comment by Gbury** [comment]
That seems interesting but there are still a few important details.
First, this prevent including regular paths that uses dots in their names, that could be a problem.
Second, in your example, using `-I M.N.P`:
- currently the `-I` option includes directories, not single files
- what is the status of M, N, P ? Is M a module name, a package name ? Are nested module currently allowed in dedukti ? Or would M and N be package names, r rather, `M.N` denote a sub-library ?
- with just this, I don't see the difference with `-I M/N/P` ; basically the feature that is currently missing in dedukti is a way for dedukti to translate what would ba package (or library) name into a path, and this can be done in several ways: either use a global dedukti path, such that package (or library) `foo` is always installed in `/global_path/foo`, or maintain a database from package names to paths.

Lastly, a separate tool becomes quite quickly heavy to use: suppose you have 3 dependencies, you would need the options `-I $(dkfind query foo1) -I $(dkfind query foo2) -I $(dkfind query foo3)`. Also, this would require to list all dependencies in the correct order including recursive dependencies, etc...

**Comment by Gbury** [comment]
For reference I don't think Coq has a separate tool for that.

**Comment by ejgallego** [comment]
> For reference I don't think Coq has a separate tool for that.

Indeed this is a big problem and has given many headaches and loss of thousands of man-hours.

The current plan for Coq is to use Dune's package format, see https://github.com/coq/coq/issues/7312 and  https://github.com/coq/coq/issues/8052

**Comment by fblanqui** [comment]
I don't understand why we would need a dkfind if we use opam: opam 
already provides these commands.


Le 04/09/2018 à 13:53, fthire a écrit :

**Comment by rafoo** [comment]
> I don't understand why we would need a dkfind if we use opam: opam already provides these commands.

The Opam commands have a very different (and complementary) behaviour from the Findlib ones. The role of Opam is to download, compile and install a package and its dependencies whereas the role of Findlib is to locate a previously installed package.

We could agree on a convention for the location of each installed library, say `$(opam config var lib)/dedukti/<library_name>`, but this is not very robust.

**Comment by fblanqui** [comment]
Why?


Le 06/09/2018 à 12:10, Raphaël Cauderlier a écrit :

**Comment by francoisthire** [comment]
What is the current plan @fblanqui for lambdapi?
We have a repository with some developments here: https://github.com/Deducteam/Libraries
What  do you think @Gaspi @GuillaumeGen ?

**Comment by fblanqui** [comment]
No plan for the moment and for the coming months... :-(. The priority is unification and tactics.

**Comment by francoisthire** [comment]
See Lambdapi.
